woocommerce/woocommerce-admin
# Fix `npm run build` resulting in non functional app.

## TLDR;

There's a bug with the default `WC_ADMIN_PHASE` value that ends up configuring the app to not have any pages registered.  This pull:

- introduces a new `build:plugin` script.
- uses that for the `build` script instead of `build:core`.

Once the `core` build is put in use, we can re-evaluate what a "production" build means, but for now the assumption is that it is equivalent to a plugin release from the standpoint of active features.

So... possible follow-ups:

- is `core` needed as a configuration option?  If not, maybe remove in a subsequent pull.
- if it is needed, then evaluate what are better defaults than everything off.

## How this was discovered (or Alice in Wonderland)

Today I decided it'd be a good idea to execute `npm run build` and verify that the pulls I did worked as expected in a production build.  This led to a good old fashioned rabbit-hole (see p1569440562154300-slack-CBKK5KM41):

- App didn't load after running `npm run build`
- I tried the usual reset trick `rm -rf node_modules` and even `npm run clean` before `npm run build`.  Still no dice.
- Pinged team in slack
- I tried completely wiping my repo and re-installing.
- I tried doing the same on tagged versions (`0.18.0`, `0.17.0`, `0.16.0`, `0.15.0`), no dice.
- I _verified_ at least that the release build (plugin zip) worked. It did 😌 
- At first @jeffstieler said it wasn't working for him, but then he said doing the `node_modules` dump ended up working for him.  But still not for me!!

So, I ended up taking a bit of a break before coming back and doing some debugging and I ended up discovering that there were no pages being registered for the router (`getPages()` was returning an empty array) which in turn led me to the Feature Flag system.  It turned out if the`WC_ADMIN_PHASE` environment variable is undefined, then it defaults to `core`.  Here's what the configuration is for the `core` phase:

```json
{
	"features": {
		"activity-panels": false,
		"analytics": false,
		"analytics-dashboard": false,
		"analytics-dashboard/customizable": false,
		"devdocs": false,
		"onboarding": false,
		"store-alerts": false
	}
}
```

🤔 seems like that means nothing will be active then right?  Right.  That was the problem.

What's puzzling to me is **how was this ever working for others**?  I'm _guessing_ it's possible folks have exported `WC_ADMIN_PHASE` to an explicit value in their environment, but apart from that 🤷‍♂

## Testing

- run `npm run build`
- verify app works!  💰 
- run `npm run dev`
- verify app works with additional features (eg `onboarding`) active. 🏃
